### PR TITLE
Rewrite chat input to use textarea and ChatGPT-style command selection (fixes inline command bugs)

### LINKED ISSUE #2357: Rewrite chat input and inline commands to ChatGPT-like style
**Is your feature request related to a problem? Please describe.**
The current command implementations cause a lot of bugs: #2326, #2264, #1890, #2032, #2087

**Describe the solution you'd like**
The input should be responsible only for handling user input (i.e. text) and should not interfere with any additional functionality.

### PR DESCRIPTION
**Pull request by AidanShipperley**
This PR aims to resolve #2357.

---

This PR rewrites the chat input to use a standard `<textarea>`, removing all `contentEditable` and inline command rendering logic. The input now only handles user text, which fixes a long-standing set of bugs around pasted content (see #2326, #2264, #1890, #2032, #2087). From my testing, I believe this solves all issues with pasted content and maintains compatibility with the existing [command APIs](https://docs.chainlit.io/concepts/command), so it should be non-breaking.

**Key changes:**
- **No more inline commands:** All selected commands are now shown as animated pill buttons below the input instead of being rendered inline as rich text along with user input.
- **ChatGPT-style `/` command menu:** Typing `/` brings up a floating, animated command menu with full keyboard and mouse support.
- **"Tools" popover:** Non-button commands are still accessible via a dropdown popover now clearly labeled as "Tools" similar to ChatGPT with smooth navigation and tooltips.
- **Robust paste handling:** Pasting text or files works as expected, with no interference from command rendering.

**Video demo:**
> https://github.com/user-attachments/assets/29573ee2-2eda-4522-8a3b-ec8507ecf7da

This should resolve all the major issues with the previous inline command implementation and make the input UX much more reliable.

Let me know if you have any feedback or want any tweaks, as I am not as experienced with frontend development.

### CODE REVIEW

**Comment by hayescode** [comment]
Thank you for your contribution @AidanShipperley ! This looks like a nice improvement and solves a lot of these whack-a-mole issues we've been seeing. Can you take a look at the CI there's an issue with the copilot e2e. Everything else looks great and I'm eager to merge this.

@asvishnyakov can you check this out please? I know you have been doing lots of work on improving this area of the code and it would be good to get your review.

**Comment by hayescode** [comment]
@AidanShipperley tested and works perfectly! Thank you for the high-quality PR!

<details>
<summary>Test Code</summary>

```py
import chainlit as cl
from chainlit.types import CommandDict

a = cl.ChatProfile(name="Assistant A", markdown_description="Assistant A")
b = cl.ChatProfile(name="Assistant B", markdown_description="Assistant B")

search = CommandDict(
    id="Search",
    icon="globe",
    description="Web Search",
    button=False,
    persistent=True,
)
image = CommandDict(
    id="Picture",
    icon="image",
    description="DALLE 3",
    button=True,
    persistent=True,
)
tool = CommandDict(
    id="tool",
    icon="sparkle",
    description="Tool",
    button=True,
    persistent=False,
)
tool2 = CommandDict(
    id="tool2",
    icon="sparkle",
    description="Tool 2",
    button=False,
    persistent=False,
)
commands = [search, image, tool, tool2]


@cl.set_chat_profiles
async def send_profiles():
    return [a, b]


@cl.on_chat_start
async def start():
    if cl.user_session.get("chat_profile") == "Assistant A":
        await cl.context.emitter.set_commands(commands)
    elif cl.user_session.get("chat_profile") == "Assistant B":
        await cl.context.emitter.set_commands([])


@cl.on_message
async def message(msg: cl.Message):
    print(f"on_message: {msg.to_dict()}")
```

</details>

I'm re-running the CI now to check if this was a transient failure. If it's still failed can you please check/fix? Once passing I'll approve/merge.

Thanks again @AidanShipperley !

**Comment by asvishnyakov** [comment]
@hayescode I definitely will! But I would appreciate your, @rajanarahul93’s, @mihidumh’s, and anyone else’s review and testing, especially for edge cases I may not be aware of

**Review by asvishnyakov — DISMISSED** [review]
@hayescode @AidanShipperley CI is failing because copilot tests depend on `contenteditable` attribute, which no longer exist in this PR. Please update them

**Comment by AidanShipperley** [comment]
Thank you for identifying the issue @asvishnyakov & @hayescode! I adjusted the copilot tests to now check for the native placeholder attribute for `textarea` instead of `contenteditable` divs with `data-placeholder` attributes.

**Review by hayescode — APPROVED** [review]

**Comment by asvishnyakov** [comment]
@AidanShipperley I'll check & test everything a bit later. Thank you for the contribution!

**Comment by rajanarahul93** [comment]
@asvishnyakov I’ve tested the changes locally and they’re working great. The textarea-based input feels much more stable, and the command selection flow is good. No issues encountered so far.

**Comment by mihidumh** [comment]
Looks good from my testing!

Not sure if it's better raising this a seperate issue, but wondering if maybe the UI should communicate that the input box is markdown, and code should be entered in code blocks etc?

For example webex does this:
image

And even this github input box has this with a link to doco:
image

**Comment by asvishnyakov** [comment]
@mihidumh Yeah, I agree it confuses users and should probably be an opt-out option, but that’s a separate issue. I’d appreciate it if you could create one.

File: cypress/e2e/command/spec.cy.ts [hunk 1]
```diff
@@ -27,6 +32,337 @@ describe('Command', () => {
...
+    // After selecting Picture, backend clears commands -> tools button disappears
+    cy.get('#command-button').should('not.exist');
+  });
+
+  it('should correctly display and interact with commands', () => {
+    cy.visit('/');
+    cy.get('#chat-input').should('exist');
+    cy.wait(500);
+
+    // Test 1: Check initial command buttons (Search and StickyButton are button commands)
```
Comment (asvishnyakov) [inline cypress/e2e/command/spec.cy.ts]: You can use [Cypress Steps](https://github.com/filiphric/cypress-plugin-steps) now

File: frontend/src/components/AutoResizeTextarea.tsx [hunk 2]
```diff
@@ -25,10 +26,10 @@ const AutoResizeTextarea = ({
     const textarea = textareaRef.current;
     if (!textarea || !onPaste) return;
 
-    textarea.addEventListener('paste', onPaste);
+    textarea.addEventListener('paste', onPaste as any);
 
     return () => {
-      textarea.removeEventListener('paste', onPaste);
+      textarea.removeEventListener('paste', onPaste as any);
```
Comment (asvishnyakov) [inline frontend/src/components/AutoResizeTextarea.tsx]: Why? Types seems already match

File: frontend/src/components/chat/MessageComposer/Input.tsx [hunk 3]
```diff
@@ -220,153 +91,230 @@ const Input = forwardRef<InputMethods, Props>(
...
+          .command-list-container::-webkit-scrollbar-thumb {
+            background-color: hsl(var(--muted-foreground) / 0.3);
+            border-radius: 2px;
+          }
+
+          .command-list-container {
+            scrollbar-width: thin;
+            scrollbar-color: hsl(var(--muted-foreground) / 0.3) transparent;
+          }
+        `}</style>
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/Input.tsx]: These animations should be in [tailwind.config.js](https://v3.tailwindcss.com/docs/animation#customizing-your-theme) to allow easy customization in projects
We [already have](https://github.com/Chainlit/chainlit/blob/e0bf2a51d9df6b0bd7c61c4417b609bac6283b57/frontend/tailwind.config.js#L66-L87) animations here

This will also allow you to reuse code from above and avoid duplication

File: frontend/src/components/chat/MessageComposer/Input.tsx [hunk 4]
```diff
@@ -220,153 +91,230 @@ const Input = forwardRef<InputMethods, Props>(
...
-            <Command className="rounded-lg border shadow-md">
-              <CommandList>
-                <CommandGroup>
+        {showCommands && filteredCommands.length > 0 && (
+          <div
+            className="absolute z-50 left-0 command-menu-animate"
+            style={{
+              bottom: '100%',
+              marginBottom: '12px'
+            }}
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/Input.tsx]: Why not `bottom-full mb-3`?

File: frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx [hunk 5]
```diff
@@ -3,94 +3,325 @@ import {
...
+                    {!hasSelectedNonButtonCommand && (
+                      <span className="overflow-hidden transition-all duration-300 opacity-100 w-auto max-w-[100px]">
+                        Tools
+                      </span>
+                    )}
+                  </Button>
+                </PopoverTrigger>
+              </TooltipTrigger>
+              <TooltipContent>
+                <p>{hasSelectedNonButtonCommand ? 'Change Tool' : 'Available Tools'}</p>
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx]: This should be put to translations and used via `Translator`

File: frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx [hunk 6]
```diff
@@ -3,94 +3,325 @@ import {
...
+        .command-list-container {
+          scrollbar-width: thin;
+          scrollbar-color: hsl(var(--muted-foreground) / 0.3) transparent;
+        }
+
+        .command-button-transition {
+          transition: width 0.3s cubic-bezier(0.34, 1.56, 0.64, 1), 
+                      padding 0.3s cubic-bezier(0.34, 1.56, 0.64, 1);
+        }
+      `}</style>
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx]: These animations should be in [tailwind.config.js](https://v3.tailwindcss.com/docs/animation#customizing-your-theme) to allow easy customization in projects
We [already have](https://github.com/Chainlit/chainlit/blob/e0bf2a51d9df6b0bd7c61c4417b609bac6283b57/frontend/tailwind.config.js#L66-L87) animations here

This will also allow you to reuse code from above and avoid duplication

File: frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx [hunk 7]
```diff
@@ -3,94 +3,325 @@ import {
...
+        buttonRef.current?.focus();
+        break;
+    }
+  };
+
+  const handleCommandSelect = (command: ICommand) => {
+    onCommandSelect(command);
+    setOpen(false);
+    cancelTooltipOpen();
+  };
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandPopoverButton.tsx]: This code seems to be copy-pasted from `Index.tsx`. Can we reuse it? May be create separate component

File: frontend/src/components/chat/MessageComposer/CommandButtons.tsx [hunk 8]
```diff
@@ -18,53 +20,179 @@ interface Props {
...
+          background-color: #0066FF;
+          border-radius: 1px;
+          animation: expand-width 0.3s cubic-bezier(0.34, 1.56, 0.64, 1);
+        }
+        
+        @keyframes expand-width {
+          0% { width: 0; }
+          100% { width: 30%; }
+        }
+      `}</style>
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandButtons.tsx]: These animations should be in [tailwind.config.js](https://v3.tailwindcss.com/docs/animation#customizing-your-theme) to allow easy customization in projects
We [already have](https://github.com/Chainlit/chainlit/blob/e0bf2a51d9df6b0bd7c61c4417b609bac6283b57/frontend/tailwind.config.js#L66-L87) animations here

This will also allow you to reuse code from above and avoid duplication

File: frontend/src/components/chat/MessageComposer/CommandButtons.tsx [hunk 9]
```diff
@@ -18,53 +20,179 @@ interface Props {
...
+            >
+              {command.id}
+            </span>
+            <div
+              className={cn(
+                "ml-1 transition-all duration-300 flex items-center",
+                isSelected ? "w-4 opacity-60" : "w-0 opacity-0"
+              )}
+            >
+              <X className="!size-4 text-[#0066FF]" />
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandButtons.tsx]: Please use one of predefined colors from theme or [define your own](https://v3.tailwindcss.com/docs/customizing-colors#using-custom-colors) in `tailwind.config.js`

File: frontend/src/components/chat/MessageComposer/CommandButtons.tsx [hunk 10]
```diff
@@ -18,53 +20,179 @@ interface Props {
...
+            isSelected && 'text-[#0066FF] command-selected',
+            isAnimating && 'animate-bounce-subtle',
+            !hasInitialized && 'opacity-0',
+            hasInitialized && 'opacity-100'
+          )}
+          onClick={handleClick}
+          style={{
+            animationDelay: hasInitialized ? '0ms' : `${index * 50}ms`,
+            '--hover-bg': 'hsl(var(--muted))',
+            '--hover-color': isSelected ? '#0066FF' : 'inherit'
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandButtons.tsx]: Please use [tailwind's utils](https://v3.tailwindcss.com/docs/hover-focus-and-other-states#hover-focus-and-active)

File: frontend/src/components/chat/MessageComposer/CommandButtons.tsx [hunk 11]
```diff
@@ -18,53 +20,179 @@ interface Props {
...
+        .command-selected::after {
+          content: '';
+          position: absolute;
+          bottom: -2px;
+          left: 50%;
+          transform: translateX(-50%);
+          width: 30%;
+          height: 2px;
+          background-color: #0066FF;
+          border-radius: 1px;
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/CommandButtons.tsx]: All of this can be done with Tailwind utils

File: frontend/src/components/chat/MessageComposer/Input.tsx [hunk 12]
```diff
@@ -220,153 +91,230 @@ const Input = forwardRef<InputMethods, Props>(
...
+
+          .command-list-container::-webkit-scrollbar-thumb {
+            background-color: hsl(var(--muted-foreground) / 0.3);
+            border-radius: 2px;
+          }
+
+          .command-list-container {
+            scrollbar-width: thin;
+            scrollbar-color: hsl(var(--muted-foreground) / 0.3) transparent;
+          }
```
Comment (asvishnyakov) [inline frontend/src/components/chat/MessageComposer/Input.tsx]: Do we really need this scrollbar (and WebKit-specific) code here? I know it looks good, but may we at least move it to frontend's `index.css` then and reuse

File: cypress/e2e/command/spec.cy.ts [hunk 13]
```diff
@@ -27,6 +32,337 @@ describe('Command', () => {
...
+
+    cy.get('#command-Search').should('exist');
+    cy.get('#command-Search').click();
+
+    cy.get('#command-Search').should('satisfy', ($el) => {
+      return (
+        $el.hasClass('command-selected') ||
+        $el.hasClass('text-[#0066FF]') ||
+        $el.find('span').hasClass('text-[#0066FF]') ||
+        $el.find('.text-\\[\\#0066FF\\]').length > 0
```
Comment (asvishnyakov) [inline cypress/e2e/command/spec.cy.ts]: If you check something like `command now disabled` here, it's better to check `:disabled` or use other attribute, even if it's introduced specifically for testing purposes. Styles may change over time

File: cypress/e2e/command/spec.cy.ts [hunk 14]
```diff
@@ -27,6 +32,337 @@ describe('Command', () => {
...
+
+  it('should handle keyboard navigation in command menu', () => {
+    cy.visit('/');
+    cy.get('#chat-input').should('exist');
+    cy.wait(500);
+
+    cy.get('#chat-input').click().clear();
+    cy.get('#chat-input').type('/');
+    cy.get('[data-index]').should('have.length.at.least', 3);
+    cy.get('[data-index="0"]').should('have.class', 'bg-accent');
```
Comment (asvishnyakov) [inline cypress/e2e/command/spec.cy.ts]: If you check something like `command now selected` here, it's better to check `:hover` or use other attribute, even if it's introduced specifically for testing purposes. Styles may change over time

**Comment by asvishnyakov** [comment]
@AidanShipperley And regardless of testing:

This seems like a bug to me: if you send the Picture command, all commands disappear:

https://github.com/user-attachments/assets/168d64b0-a388-4dcc-aacc-031ca909664b


Let me know if it doesnt’t, as I’m not very familiar with the specifics of commands. At first, I also thought that command deselection was a bug, until I found out about the `persistent` attribute :)

Everything else seems fine to me.

---

To be honest, I don’t like these animations, especially the one that enlarges a command when you select it, as they feel too heavy and unnatural for this UI. However, that’s more of a personal preference, so you may keep them all if you want.

**Comment by hayescode** [comment]
@AidanShipperley I want to include your PR here in the next release, can you address @asvishnyakov 's comments?

**Comment by AidanShipperley** [comment]
> @AidanShipperley I want to include your PR here in the next release, can you address @asvishnyakov 's comments?

@hayescode Yeah, I can take a look at them today!

**Comment by hayescode** [comment]
Thank you @AidanShipperley !

**Comment by hayescode** [comment]
@AidanShipperley thank you again for this, major improvement! One thing i noticed in 3.7.1.1 is that the Tools button is now not wide enough. I think it looks great before when it said "Tools". Would you mind making a PR to fix this?

image

**Comment by susanaryal** [comment]
@AidanShipperley, this is such a great feature, and thank you for this!

Just wanted to share the Copilot mode in 3.7.1.1 — I noticed that when tools have long names they can overflow and visually crowd the interface, as shown in the attached screenshot.

If you change the label's three dots `...` to `Tools` and also if I had audio mode enabled which would have one icon for audio, it would occupy even more width in Copilot mode.
Screenshot 2025-08-21 at 4 28 48 PM

It would be great if the tools names hid and icon only displayed automatically based on width especially in the Copilot mode to keep the layout clean and user-friendly? 

I was able to hide the tools name and display icons only somehow by zooming in the browser. But if we could achive this same behavior in the previous screenshot's zoom level as well.
Screenshot 2025-08-21 at 4 29 10 PM

Thanks again for all the thoughtful work — really appreciate it!
